device/common/include/traccc/finding/device/impl/add_links_for_holes.ipp

```diff
@@ -0,0 +1,63 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+namespace traccc::device {
+
+TRACCC_DEVICE inline void add_links_for_holes(
+    std::size_t globalIndex,
+    vecmem::data::vector_view<const unsigned int> n_candidates_view,
+    bound_track_parameters_collection_types::const_view in_params_view,
+    const unsigned int step, const unsigned int& n_max_candidates,
+    bound_track_parameters_collection_types::view out_params_view,
+    vecmem::data::vector_view<candidate_link> links_view,
+    unsigned int& n_total_candidates) {
+
+    // Number of candidates per parameter
+    vecmem::device_vector<const unsigned int> n_candidates(n_candidates_view);
+
+    if (globalIndex >= n_candidates.size()) {
+        return;
+    }
+
+    // Input parameters
+    bound_track_parameters_collection_types::const_device in_params(
+        in_params_view);
+
+    // Output parameters
+    bound_track_parameters_collection_types::device out_params(out_params_view);
+
+    // Links
+    vecmem::device_vector<candidate_link> links(links_view);
+
+    // Last step ID
+    const unsigned int previous_step =
+        (step == 0) ? std::numeric_limits<unsigned int>::max() : step - 1;
+
+    if (n_candidates[globalIndex] == 0u) {
+
+        // Add measurement candidates to link
+        vecmem::device_atomic_ref<unsigned int> num_total_candidates(
+            n_total_candidates);
+
+        const unsigned int l_pos = num_total_candidates.fetch_add(1);
+
+        if (l_pos >= n_max_candidates) {
+
+            n_total_candidates = n_max_candidates;
+            return;
+        }
+
+        // Add a dummy link
+        links[l_pos] = {{previous_step, globalIndex},
+                        std::numeric_limits<unsigned int>::max()};
+        out_params[l_pos] = in_params[globalIndex];
```

Comment:
Same here. Using `.at(...)` in all these calls could make debugging a little easier in the future.

Reply:
OK